File: RentSystem/src/main/java/com/rentSystem/project/NotificationModule.java
package com.rentSystem.project;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class NotificationModule implements Notifier {

    private final Map<String, List<Observer>> observers = new HashMap<>();

    @Override
    public void subscribe(Observer observer, String notificationType) {
        observers.computeIfAbsent(notificationType, k -> new ArrayList<>()).add(observer);
        System.out.println("Observer subscribed to " + notificationType);
    }

    @Override
    public void unsubscribe(Observer observer, String notificationType) {
        List<Observer> userObservers = observers.get(notificationType);
        if (userObservers != null) {
            userObservers.remove(observer);
            System.out.println("Observer unsubscribed from " + notificationType);
        }
    }

    @Override
    public void notifyObservers(Notification notification) {
        List<Observer> userObservers = observers.get(notification.getNotificationType());
        if (userObservers != null) {
            for (Observer observer : userObservers) {
                observer.update(notification);
            }
        }
    }
}

File: RentSystem/src/main/kotlin/com/rentSystem/project/NotificationModule.kt
package com.rentSystem.project

class NotificationModule : Notifier {
    private val observers = mutableMapOf<String, MutableList<Observer>>() // Карта: тип уведомления -> список наблюдателей

    override fun subscribe(observer: Observer, notificationType: String) {
        observers.computeIfAbsent(notificationType) { mutableListOf() }.add(observer)
        println("${observer::class.simpleName} с ID ${observer.id} подписан на уведомления типа: $notificationType")
    }

    override fun unsubscribe(observer: Observer, notificationType: String) {
        observers[notificationType]?.remove(observer)
        println("${observer::class.simpleName} с ID ${observer.id} отписан от уведомлений типа: $notificationType")
    }

    override fun notifyObservers(notification: Notification) {
        val observersForType = observers[notification.notificationType] ?: emptyList()
        observersForType.forEach { it.update(notification) }
        println("Уведомление отправлено: ${notification.messageText}")
    }
}


